sdk/dotnet/Pulumi/Core/OutputWatcher.cs

```diff
@@ -0,0 +1,68 @@
+// Copyright 2016-2020, Pulumi Corporation
+
+using System;
+using System.Collections.Immutable;
+using System.Threading.Tasks;
+
+namespace Pulumi
+{
+    /// <summary>
+    /// Allows extracting some internal insights about an instance of
+    /// <see cref="Output{T}"/>. 
+    /// </summary>
+    public sealed class OutputWatcher<T>
+    {
+        private readonly Output<T> _output;
+
+        public OutputWatcher(Output<T> output)
+        {
+            _output = output;
+            _output.DataTask.ContinueWith(t =>
+            {
+                var data = t.Result;
+                var args = new OutputResolvedArgs<T>(data.Value, data.IsKnown, data.IsSecret);
+                Resolved?.Invoke(this, args);
+            });
+        }
+
+        /// <summary>
+        /// Retrieve a set of resources that the output depends on.
+        /// </summary>
+        /// <returns></returns>
+        public Task<ImmutableHashSet<Resource>> GetDependenciesAsync()
+            => (_output as IOutput).GetResourcesAsync();
+
+        /// <summary>
+        /// Fires when the output value is resolved.
+        /// </summary>
+        public event EventHandler<OutputResolvedArgs<T>>? Resolved;
```

Comment:
do we need this? 

Couldn't this just be `OutputUtilities.GetIsKnown(Output)`?  It would return a `Task<bool>` which anyone could use as they see fit?

Reply:
I thought it makes sense to prevent people from avaiting this task and blocking their programs?

Reply:
1. that's a good point.  noodling on it.
2. i'm skeptical that that is a problem, esp. if we hide this behind an OutputUtilities class and doc it as something to be very careful of.  
3. even if this is the right approach, the above code won't work.  it's fundamentally racey on .net.  user would have to write:

```c#
var watcher = new Watcher(someOutput);
watcher.Resolved += ...
```

However, it *will* happen that `.Resolved` will be called prior to the event hookup because .net will happily execute the `.ContinueWith` in teh constructor immediately.

If we want to go this route, a better approach is more like: `OutputUtilities.WhenIsKnown(Output op, Action<bool> work)`.   That can just attach the action to the DataTask directly.

Reply:
Great point on (3)! 2 makes sense given your remark about not exposing `Value`, otherwise I would be more skeptical.